File: qc-application/src/main/java/com/eager/questioncloud/application/subscribe/SubscribeProcessor.java
package com.eager.questioncloud.application.subscribe;

import com.eager.questioncloud.domain.creator.CreatorRepository;
import com.eager.questioncloud.domain.subscribe.Subscribe;
import com.eager.questioncloud.domain.subscribe.SubscribeRepository;
import com.eager.questioncloud.exception.CustomException;
import com.eager.questioncloud.exception.Error;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Component
@RequiredArgsConstructor
public class SubscribeProcessor {
    private final SubscribeRepository subscribeRepository;
    private final CreatorRepository creatorRepository;

    public Subscribe subscribe(Long userId, Long creatorId) {
        if (!isActiveCreator(creatorId)) {
            throw new CustomException(Error.NOT_FOUND);
        }

        if (isAlreadySubscribed(userId, creatorId)) {
            throw new CustomException(Error.ALREADY_SUBSCRIBE_CREATOR);
        }

        return subscribeRepository.save(Subscribe.create(userId, creatorId));
    }

    @Transactional
    public void unSubscribe(Long userId, Long creatorId) {
        subscribeRepository.unSubscribe(userId, creatorId);
    }

    private Boolean isActiveCreator(Long creatorId) {
        return creatorRepository.existsById(creatorId);
    }

    private Boolean isAlreadySubscribed(Long userId, Long creatorId) {
        return subscribeRepository.isSubscribed(userId, creatorId);
    }
}

File: qc-domain/qc-subscribe/qc-subscribe-api/src/main/kotlin/com/eager/questioncloud/subscribe/implement/SubscribeProcessor.kt
package com.eager.questioncloud.subscribe.implement

import com.eager.questioncloud.common.exception.CoreException
import com.eager.questioncloud.common.exception.Error
import com.eager.questioncloud.creator.api.internal.CreatorQueryAPI
import com.eager.questioncloud.subscribe.domain.Subscribe
import com.eager.questioncloud.subscribe.infrastructure.repository.SubscribeRepository
import org.springframework.stereotype.Component

@Component
class SubscribeProcessor(
    private val subscribeRepository: SubscribeRepository,
    private val creatorQueryAPI: CreatorQueryAPI,
) {
    fun subscribe(userId: Long, creatorId: Long) {
        if (!isActiveCreator(creatorId)) {
            throw CoreException(Error.NOT_FOUND)
        }

        if (isAlreadySubscribed(userId, creatorId)) {
            throw CoreException(Error.ALREADY_SUBSCRIBE_CREATOR)
        }

        subscribeRepository.save(Subscribe.create(userId, creatorId))
    }

    fun unSubscribe(userId: Long, creatorId: Long) {
        subscribeRepository.unSubscribe(userId, creatorId)
    }

    private fun isActiveCreator(creatorId: Long): Boolean {
        return creatorQueryAPI.isExistsById(creatorId)
    }

    private fun isAlreadySubscribed(userId: Long, creatorId: Long): Boolean {
        return subscribeRepository.isSubscribed(userId, creatorId)
    }
}
